// @netlify/sdk/dist/cli/index.js
import { resolve } from "path";
import { execSync } from "child_process";
import chalk from "chalk";
import pkg from "fs-extra";
import { outdent as javascript } from "outdent";
import { build, context } from "esbuild";
import { getPackageVersion } from "../integration/ui/utils/package-version.js";
import { handler } from "../constants/handler.js";
const { ensureFile, ensureDir, copySync, copyFile, copyFileSync, writeFile, writeJSON, existsSync, readJSON, } = pkg;
export { generateConnectPlugin, createDevelopmentContentEngine, } from "../integration/connect/cli.js";
export const generateBuildPlugin = async (integration, config, opts) => {
    if (Object.values(integration.buildHooks).every((value) => !value)) {
        console.log(chalk.white("No build hooks found."));
        return;
    }
    const { outDir, path } = opts;
    const integrationPackageJson = await readJSON(resolve(path, "package.json"));
    const { dependencies } = integrationPackageJson;
    await ensureDir(resolve(outDir, "build"));
    await copyFile(resolve(outDir, "index.js"), resolve(outDir, "build/integration.js"));
    let exportedPluginCode = javascript `
    import { integration } from './integration.js';
    integration._slug = '${config.slug}';\n
  `;
    exportedPluginCode += javascript `
    export default function netlifyIntegration(_inputs, { netlifyConfig }) {
      return integration.getBuildEventHandlersToRun({ netlifyConfig });
    };`;
    // generate random 5 character string
    const randomString = Math.random().toString(36).substring(2, 7);
    const version = `0.0.0-${randomString}`;
    const name = `${config.slug}-buildhooks`;
    const sdkVersion = getPackageVersion();
    if (existsSync(resolve(path, "assets"))) {
        copySync(resolve(path, "assets"), resolve(outDir, "build/assets"));
    }
    await writeFile(resolve(outDir, "build/index.js"), exportedPluginCode);
    await writeJSON(resolve(outDir, "build/package.json"), {
        main: "index.js",
        type: "module",
        version,
        name,
        dependencies: { ...dependencies, "@netlify/sdk": sdkVersion },
    });
    await writeFile(resolve(outDir, "build/manifest.yml"), `name: ${name}`);
    execSync(`npm pack`, {
        cwd: resolve(outDir, "build"),
    });
    const packedFile = `${name
        .replaceAll("@", "")
        .replaceAll("/", "-")}-${version}.tgz`;
    await ensureDir(resolve(outDir, "site/static/packages"));
    copyFileSync(resolve(outDir, `build/${packedFile}`), resolve(outDir, "site/static/packages", "buildhooks.tgz"));
};
export const createIntegrationUIFiles = async (surfaceScriptUrl, outDir, slug) => {
    // You're probably thinking, "Why are you using a the date as a query param?"
    // Well, it's because we want to bust the cache each time the rebuild happens.
    // It turns out functions that run on the ESBuild plugins import the file once
    // and then cache it. So, we need to bust the cache each time we rebuild.
    const now = Date.now();
    const uiModule = await import(`${outDir}/site/static/ui/integration-ui.js?${now}`);
    const { config, script } = uiModule.integrationUI.generateUIConfig(surfaceScriptUrl);
    config.integrationSlug = slug;
    console.log(chalk.white("Generating UI surfaces"));
    await ensureDir(resolve(outDir, "site/static/ui"));
    await writeFile(resolve(outDir, "site/static/ui/integration-ui.json"), JSON.stringify(config));
    if (script) {
        const builtPath = resolve(outDir, "site/static/ui/integration-ui.js");
        await ensureFile(builtPath);
        await writeFile(builtPath, script, { flag: "a" });
    }
};
export const buildIntegrationUI = async ({ outDir, mainPath, surfaceScriptUrl, watch, id, }) => {
    let integrationUIFilePath;
    if (existsSync(resolve(mainPath, "../ui/index.ts"))) {
        integrationUIFilePath = resolve(mainPath, "../ui/index.ts");
    }
    else if (existsSync(resolve(mainPath, "../ui/index.js"))) {
        integrationUIFilePath = resolve(mainPath, "../ui/index.js");
    }
    else {
        console.log(chalk.white("No integration UI found"));
        return;
    }
    if (integrationUIFilePath) {
        console.log(chalk.white("Building integration UI..."));
        const buildOptions = {
            entryPoints: [integrationUIFilePath],
            bundle: true,
            format: "esm",
            outfile: `${outDir}/site/static/ui/integration-ui.js`,
            platform: "node",
            keepNames: true,
            minify: true,
        };
        if (watch) {
            try {
                const buildContext = await context({
                    ...buildOptions,
                    plugins: [
                        {
                            name: "rebuild-listener",
                            setup(build) {
                                let shouldNotifyRebuildListeners = false;
                                // sometimes esbuild immediately rebuilds twice on start,
                                setTimeout(() => 
                                // so only notify listeners if it's been more than 100ms since the build started
                                (shouldNotifyRebuildListeners = true), 100);
                                let notifying = false;
                                build.onEnd(async () => {
                                    if (notifying) {
                                        return;
                                    }
                                    notifying = true;
                                    if (shouldNotifyRebuildListeners) {
                                        await createIntegrationUIFiles(surfaceScriptUrl, outDir, id);
                                        console.log(chalk.green("Integration UI rebuilt"));
                                    }
                                    notifying = false;
                                });
                            },
                        },
                    ],
                });
                await buildContext.watch();
                await buildContext.rebuild();
            }
            catch (e) {
                console.log(chalk.red("Could not build integration UI"));
                console.log(e);
                throw e;
            }
        }
        else {
            try {
                await build(buildOptions);
            }
            catch (e) {
                console.log(chalk.red("Could not build integration UI"));
                console.log(e);
                throw e;
            }
        }
        await createIntegrationUIFiles(surfaceScriptUrl, outDir, id);
    }
};
export const processHandlers = async (opts, config) => {
    const { outDir } = opts;
    const handlerDir = resolve(outDir, "site/netlify/functions");
    await ensureDir(handlerDir);
    await ensureDir(resolve(outDir, "site/netlify/integration"));
    await copyFile(resolve(outDir, "index.js"), resolve(outDir, "site/netlify/integration/index.js"));
    const handlerCode = handler("../integration", config);
    await writeFile(resolve(handlerDir, "handler.js"), handlerCode);
};
export const processWrappers = async (wrappers, opts) => {
    const { outDir } = opts;
    console.log(chalk.white("Generating wrapper functions..."));
    const wrapperFile = resolve(outDir, "wrappers/index.js");
    await ensureFile(wrapperFile);
    await copyFile(resolve(outDir, "index.js"), resolve(outDir, "wrappers/integration.js"));
    let wrapperCode = javascript `import { integration } from './integration.js';\n\n`;
    Object.entries(wrappers).map(async ([key]) => {
        wrapperCode += javascript `export const ${key} = integration.wrappers['${key}'];\n`;
    });
    await writeFile(wrapperFile, wrapperCode);
};
